### PR TITLE
feat: integrate asyncapi-validator to validate incomming messages

### LINKED ISSUE #8: Integrate asyncapi-validator in Nodejs templates
@derberg commented on [Mon Mar 02 2020](https://github.com/asyncapi/generator/issues/234)

Reason:
Current Node templates nicely show what you can get from AsyncAPI file. You can generate an application that can work with Kafka or RabbitMQ, or others. When I play with those, there is one thing missing to make the templates nice, the message validation. Now you can send any message to the channel used by your application. Would be awesome to show that thanks to AsyncAPI with support of asyncapi-validator you can change that.

Description:
- Integrate https://www.npmjs.com/package/asyncapi-validator with the template
- Update docs and scratch note for release notes
- I'm wondering if it should be enabled by default or behind a parameter?

Something interesting for you @WaleedAshraf 😉 

---

@WaleedAshraf commented on [Mon Mar 02 2020](https://github.com/asyncapi/generator/issues/234#issuecomment-593330572)

🎉 Niceeee
I'd love to work on this, will try to join the next meeting and we can go from there. 👍 

---

@derberg commented on [Mon Mar 02 2020](https://github.com/asyncapi/generator/issues/234#issuecomment-593360366)

See ya tomorrow then!

### PR DESCRIPTION
**Pull request by KhudaDad414**
**Description**

- integrates asyncapi-validator into this template.

**Related issue(s)**
`Resolves #8 `

### CODE REVIEW

**Comment by KhudaDad414** [comment]
@derberg I've taken a totally different approach here. what do you think?

File: template/src/lib/message-validator.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+const AsyncApiValidator = require('asyncapi-validator');
+
+module.exports.validateMessage = async (payload, topic,name,operation)=> {
```
Comment (derberg) [inline template/src/lib/message-validator.js:3]: change `topic` to `channel` please and add some jsdoc

File: template/src/lib/message-validator.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+  let va = await AsyncApiValidator.fromSource('./asyncapi.yaml', {msgIdentifier: 'name'});
+  va.validate(name, payload, topic, operation);
```
Comment (derberg) [inline template/src/lib/message-validator.js:5]: please catch an error and format a log message nicely, like some custom text "message rejected" + the reason from error message without the rest of stack trace

**Review by derberg — COMMENTED** [review]
Looks great mate 🚀 
I only have few improvements and fix request but from the concept point of view -> well done

@WaleedAshraf wanna have a look?

@reselbob you asked me about this one a week ago, so I guess you'll be happy it is almost there. Feel free to suggest if you would expect something more. Can be done here or in a separate Pr, depending on complexity

**Comment by KhudaDad414** [comment]
@derberg made the suggested changes.

File: template/src/api/middlewares/error-logger.js [hunk 3]
```diff
@@ -1,7 +1,11 @@
 const { red, gray } = require('colors/safe');
 
 module.exports = (err, message, next) => {
-  console.error(red(`❗  ${err.message}`));
-  if (err.stack) console.error(gray(err.stack.substr(err.stack.indexOf('\n')+1)));
+  if(err.name === 'AsyncAPIValidationError'){
```
Comment (WaleedAshraf) [inline template/src/api/middlewares/error-logger.js:4]: indent seems broken 👀

**Review by WaleedAshraf — APPROVED** [review]
Looks good.

File: template/src/api/middlewares/error-logger.js (hunk 3 above)
Comment (KhudaDad414) [reply template/src/api/middlewares/error-logger.js]: corrected. (I think)

**Review by derberg — COMMENTED** [review]
one more thing, sorry but I think it would be super cool to have it in the readme

please change this
```
#publish your message.
mqtt pub -t 'smartylighting/streetlights/1/0/event/{streetlightId}/lighting/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'

# You should see the sent message in the logs of the previously started server.
```

to something like 
```

```
#publish an invalid message.
mqtt pub -t 'smartylighting/streetlights/1/0/event/123/lighting/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": "3", "sentAt": "2017-06-07T12:34:32.000Z"}'

#publish a valid message
mqtt pub -t 'smartylighting/streetlights/1/0/event/123/lighting/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'

#You should see the sent message in the logs of the previously started server.
#Notice that the server automatically validates incoming messages and logs out validation errors
```


what do you think?

**Comment by derberg** [comment]
@KhudaDad414 I would make code suggestion so you can just commit but unfortunately GH doesn't allow this to lines that were not "touched" by the PR

**Comment by KhudaDad414** [comment]
@derberg Sorry, I was out of town yesterday. pushed the suggested changes.

**Review by derberg — APPROVED** [review]
approving with pleasure

**Pull request merged**